test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Audit/AuditTests.cs

```diff
@@ -380,6 +380,139 @@ public async Task GivenAResource_WhenNotAuthorized_ThenAuditLogEntriesShouldBeCr
                 expectedAppId: TestApplications.NativeClient.ClientId);
         }
 
+        [Fact]
+        [HttpIntegrationFixtureArgumentSets(dataStores: DataStore.SqlServer)]
+        [Trait(Traits.Category, Categories.Transaction)]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenATransactionBundleWithValidEntries_WhenSuccessfulPost_ThenAuditLogEntriesShouldBeCreated()
+        {
+            if (!_fixture.IsUsingInProcTestServer)
+            {
+                // This test only works with the in-proc server with customized middleware pipeline
+                return;
+            }
+
+            // Even enteries are audit executed entry and odd entries are audit executing entry
+            List<(string expectedActions, string expectedPathSegments, HttpStatusCode? expectedStatusCodes, ResourceType? resourceType)> expectedList = new List<(string, string, HttpStatusCode?, ResourceType?)>
+            {
+                ("transaction", string.Empty, null, null),
+                ("create", "Patient", null, null),
+                ("create", "Patient", HttpStatusCode.Created, ResourceType.Patient),
+                ("create", "Patient", null, null),
+                ("create", "Patient", HttpStatusCode.OK, null),
+                ("update", "Patient/123", null, null),
+                ("update", "Patient/123", HttpStatusCode.OK, ResourceType.Patient),
+                ("update", "Patient?identifier=http:/example.org/fhir/ids|456456", null, null),
+                ("update", "Patient?identifier=http:/example.org/fhir/ids|456456", HttpStatusCode.OK, ResourceType.Patient),
+                ("transaction", string.Empty, HttpStatusCode.OK, ResourceType.Bundle),
+            };
+
+            var requestBundle = Samples.GetJsonSample("Bundle-TransactionWithValidBundleEntry");
+
+            await ExecuteAndValidateTransaction(
+               () => _client.PostBundleAsync(requestBundle.ToPoco<Hl7.Fhir.Model.Bundle>()),
+               expectedList);
+        }
+
+        [Fact]
+        [HttpIntegrationFixtureArgumentSets(dataStores: DataStore.SqlServer)]
+        [Trait(Traits.Category, Categories.Transaction)]
+        [Trait(Traits.Priority, Priority.One)]
+        public async Task GivenATransactionBundleWith_WhenAUnsuccessfulPost_ThenTransactionShouldRollBackAndAuditLogEntriesShouldBeCreated()
+        {
+            if (!_fixture.IsUsingInProcTestServer)
+            {
+                // This test only works with the in-proc server with customized middleware pipeline
+                return;
+            }
+
+            List<(string expectedActions, string expectedPathSegments, HttpStatusCode? expectedStatusCodes, ResourceType? resourceType)> expectedList = new List<(string, string, HttpStatusCode?, ResourceType?)>
+            {
+                ("transaction", string.Empty, null, null),
```

Comment:
The corresponding executing and executed rows can be merged together. For reference you might take a look #764